### DISCUSSION TITLE
Action trigger when PR base branch is updated?

### DISCUSSION THREAD

**Discussion by brianlove** [community/community#64119] [Actions]
### Select Topic Area

Product Feedback

### Body

My team has a workflow that checks whether a branch is far enough out of date that it should be rebased (this is to work around the lack of support for [semi-linear histories](https://github.com/orgs/community/discussions/8940)).  Currently the workflow runs on `pull_request` and `pull_request_review` events, which leads to the bot re-running when the PR's head branch gets updated.  However, if the base (`main`) branch gets updated, the bot does not get re-triggered unless there is a later change to the head branch.

1. Initial state - rebase bot runs and reports that things are okay
```
A --> B (main)
       \
        P (dev-branch)
```

2. Changes to `main` - no update from the bot
```
A --> B --> C (main)
       \
        P (dev-branch)
```

3. Update to `dev-branch` - bot runs and reports that `dev-branch` is one commit behind `main`
```
A --> B --> C (main)
       \
        P --> Q (dev-branch)
```

I was hoping to resolve this by finding another run trigger for the workflow, but I haven't had any luck (at least so far).  The description for [`synchronize`](https://docs.github.com/en/webhooks-and-events/webhooks/webhook-events-and-payloads?actionType=synchronize#pull_request) ("_A pull request's head branch was updated. For example, the head branch was updated from the base branch or new commits were pushed to the head branch._") is close, but I'm looking for a version that triggers on the _base_ branch, not the _head_ branch.

It would be really helpful if this additional trigger (`synchronize_base`?  `update_base`?) could be added to GitHub Actions.

**Comment by vacisdev** [comment]
We understand your need for a workflow trigger on updates to the base branch. Unfortunately, as of now, GitHub Actions does not natively support this specific trigger. However, you could consider using a scheduled workflow or a webhook to periodically check for changes to the base branch and trigger the necessary actions.

**Comment by ruudk** [comment]
I would like to have this functionality too.

**Comment by dariocc** [comment]
The `edited` action type will trigger on base branches changes (https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=edited#pull_request) but it is not a great solution because it will also trigger on PR body changes, which is weird because the kind of things you'd automate on base changes are very different from the things you'd do on PR description changes.

**Reply by sihil** [reply]
Does this fire when the specified base branch ref is changed (i.e. the PR is changed so it's base branch is changed from `main` to `feature-5`) or when further commits are pushed to the already specified branch? (Which I think is what the OP is getting at?)

**Comment by BoscoDomingo** [comment]
The fact that this hasn't yet been done is really annoying, especially for stacked workflows where the base branch is constantly updated since each PR builds on the previous one, and only one actually points to the trunk...

**Reply by klingenm** [reply]
Stacked PRs would not be helped by this.

If you have the state:
```
A --> B --> C (main)
       \
        P --> Q (dev-branch) PR1 -> base: main
               \
                Z (dev-branch2) PR2 -> base: dev-branch
```

when more commits are pushed to `main`, PR1 could get a notification and rebuild, but that build is done on the following tree:

```
A --> B --> C (main) ------\
       \                    \
        P --> Q (dev-branch) * refs/pull/1/merge
```
`refs/pull/1/merge` which is "`main` merged into `dev-branch`", is what is getting checked out in your workflow.

Then if you get a new build on PR2, which has `dev-branch` as base:
```
A --> B --> C (main)
       \
        P --> Q (dev-branch) --\
               \                \
                Z (dev-branch2) -* refs/pull/2/merge
```
it will not get the `C` commit in its build, since PR2 does not know about PR1 but only `dev-branch` how can it know?

What you would like to request is actually to set a PR base to another PR and not a branch.

I usually solve this use-case by frequently rebasing the "last" PR in the stack onto main, with the git option `rebase.updaterefs=true` and `rerere.enabled=true` making it quite painless. This only works fine when you are alone working on those branches.

**Reply by BoscoDomingo** [reply]
What I need is a way to not have branches fall out of date without a manual rebase.

In your example, the moment `main` gets new changes, I would expect the `dev-branch` PR to receive a `base-updated` event we can use to run an automated GitHub Action that rebases via the corresponding CLI command, which in turn would cause `dev-branch2` to also trigger said workflow and so on. This would rebase every single stack onto the latest `main` without any human interaction (save for conflicts, which would stop the waterfall).

**Comment by itanex** [comment]
I had a similar issue, where I had read the documentation for the `type` of `edited` and inferred that "The title or body of a pull request was edited, or the base branch of a pull request was changed" implied that meant that commits to the base branch affected this. However, as the OP question suggests a synchronized type.

The actual type needed to trigger when a commit is added to the "head (source) branch" of the PR is `synchronize`. [Source](https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=synchronize#pull_request)

**Comment by sgtsquiggs** [comment]
Using `edited` doesn't work because it will trigger on title/desc edits. We thought about skipping when the edit was only title/desc but then you end up with a passing PR because everything was skipped.

**Reply by wtfowler** [reply]
+1... in my case the Jira app is constantly editing PR bodies, triggering an `edited` event just after the `opened` event. If I conditionally skipped all the jobs in the workflow on this edit event, then the [branch protections would consider all of those successful](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/troubleshooting-required-status-checks#handling-skipped-but-required-checks) as you mentioned, and so no required status checks would actually need to pass!

**Reply by henriksquareone** [reply]
Same issue, I feel like that could potentially be a security concern as it allows bypassing branch protection rules.. would love to see a solution for this that doesn't require triggering full builds for each PR edit..

**Comment by akhilreddyt** [comment]
Using `edited` works but needs an additional step.
Use `gh api /repos/$GITHUB_REPOSITORY/issues/$PR_NUMBER/events` to get the timeline of events/changes to the PR. Use jq to find for `event` = `base_ref_changed` which indicates the base ref for the PR has been changed. Its tricky to handle though as `base_ref_changed` event will always be in the log of events once the base ref changes, so need to work with the timestamps

**Comment by infinisil** [comment]
Duplicate of https://github.com/orgs/community/discussions/35058

**Reply by 0rvar** [reply]
No it isn't

**Comment by jmaixl** [comment]
+1 this would be great especially when switching base branches that have different required checks that should trigger. Otherwise, users must run checks themselves through other interactions on pull requests which is not intuitive.

**Comment by tnielens** [comment]
A workaround some might find useful.
In the branch protection rules, a rule can be set to prevent PR merges if the base branch contains commits that are not part of the PR head branch.
https://stackoverflow.com/questions/67868483/how-to-rerun-github-actions-workflow-when-the-pull-request-base-branch-updated
